{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "54ecdb6a-49e5-4a6f-8a51-ee78386f49e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import random\n",
    "from cycler import cycler\n",
    "from matplotlib_venn import venn3  \n",
    "from matplotlib_venn.layout.venn3 import DefaultLayoutAlgorithm\n",
    "\n",
    "import gpplot as gpp\n",
    "gpp.set_aesthetics(context = 'paper')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8a562c9-bd09-4345-af01-7d8f9843c00c",
   "metadata": {},
   "source": [
    "Combining first set of CRISPick runs to assemble Jacquere in order to ...\n",
    "- Assess the guides selected from the picking scheme, troubleshooting any odd patterns observed\n",
    "- Assess if 4 guides per gene is necessary, or if this improved picking scheme enables a more compact library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5046b738-d2a9-4ae7-8730-00f7859f9a11",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/nm/6_wl7tx50tq5bqfp8vzf4m6r0000gq/T/ipykernel_92127/3886464560.py:5: DtypeWarning: Columns (27,29,30,31) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  ensembl_designs=pd.read_table(\"../Data/ensembl_113_protein_coding_aggcfd_targetlocal-sgrna-designs.txt\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input</th>\n",
       "      <th>Quota</th>\n",
       "      <th>Target Taxon</th>\n",
       "      <th>Target Gene ID</th>\n",
       "      <th>Target Gene Symbol</th>\n",
       "      <th>Target Transcript</th>\n",
       "      <th>Target Reference Coords</th>\n",
       "      <th>Target Alias</th>\n",
       "      <th>CRISPR Mechanism</th>\n",
       "      <th>Target Domain</th>\n",
       "      <th>...</th>\n",
       "      <th>On-Target Efficacy Score</th>\n",
       "      <th>On-Target Rank</th>\n",
       "      <th>Preselected As</th>\n",
       "      <th>Matching Active Arrayed Oligos</th>\n",
       "      <th>Matching Arrayed Constructs</th>\n",
       "      <th>Pools Containing Matching Construct</th>\n",
       "      <th>Pick Order</th>\n",
       "      <th>Picking Round</th>\n",
       "      <th>Picking Notes</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>4</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CRISPRko</td>\n",
       "      <td>CDS</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0210</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CP1889, CP2115</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>4</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CRISPRko</td>\n",
       "      <td>CDS</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9891</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BRDN0003487346</td>\n",
       "      <td>CP1889, CP2114, CP2115</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>4</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CRISPRko</td>\n",
       "      <td>CDS</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9622</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CP1889, CP2114, CP2115</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>4</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CRISPRko</td>\n",
       "      <td>CDS</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6634</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>4</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>ENST00000373031.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CRISPRko</td>\n",
       "      <td>CDS</td>\n",
       "      <td>...</td>\n",
       "      <td>1.1120</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GENCODE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81674</th>\n",
       "      <td>chrX:-:140091677-140092911</td>\n",
       "      <td>4</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000023.11:-:140091677-140092911</td>\n",
       "      <td>chrX:-:140091677-140092911</td>\n",
       "      <td>CRISPRko</td>\n",
       "      <td>CDS</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4118</td>\n",
       "      <td>31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHESS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81675</th>\n",
       "      <td>chrY:-:1391992-1392341;1395129-1395548;1396371...</td>\n",
       "      <td>4</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000024.10:-:1391992-1392341;1395129-1395548...</td>\n",
       "      <td>chrY:-:1391992-1392341;1395129-1395548;1396371...</td>\n",
       "      <td>CRISPRko</td>\n",
       "      <td>CDS</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9535</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Previously skipped due to: Off-target CFD100 m...</td>\n",
       "      <td>CHESS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81676</th>\n",
       "      <td>chrY:-:1391992-1392341;1395129-1395548;1396371...</td>\n",
       "      <td>4</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000024.10:-:1391992-1392341;1395129-1395548...</td>\n",
       "      <td>chrY:-:1391992-1392341;1395129-1395548;1396371...</td>\n",
       "      <td>CRISPRko</td>\n",
       "      <td>CDS</td>\n",
       "      <td>...</td>\n",
       "      <td>0.8926</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>Previously skipped due to: Off-target CFD100 m...</td>\n",
       "      <td>CHESS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81677</th>\n",
       "      <td>chrY:-:1391992-1392341;1395129-1395548;1396371...</td>\n",
       "      <td>4</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000024.10:-:1391992-1392341;1395129-1395548...</td>\n",
       "      <td>chrY:-:1391992-1392341;1395129-1395548;1396371...</td>\n",
       "      <td>CRISPRko</td>\n",
       "      <td>CDS</td>\n",
       "      <td>...</td>\n",
       "      <td>0.8820</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>Previously skipped due to: Off-target CFD100 m...</td>\n",
       "      <td>CHESS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81678</th>\n",
       "      <td>chrY:-:1391992-1392341;1395129-1395548;1396371...</td>\n",
       "      <td>4</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000024.10:-:1391992-1392341;1395129-1395548...</td>\n",
       "      <td>chrY:-:1391992-1392341;1395129-1395548;1396371...</td>\n",
       "      <td>CRISPRko</td>\n",
       "      <td>CDS</td>\n",
       "      <td>...</td>\n",
       "      <td>0.7691</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>Previously skipped due to: Off-target CFD100 m...</td>\n",
       "      <td>CHESS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81679 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Input  Quota  Target Taxon  \\\n",
       "0                                        ENSG00000000003      4          9606   \n",
       "1                                        ENSG00000000003      4          9606   \n",
       "2                                        ENSG00000000003      4          9606   \n",
       "3                                        ENSG00000000003      4          9606   \n",
       "4                                        ENSG00000000005      4          9606   \n",
       "...                                                  ...    ...           ...   \n",
       "81674                         chrX:-:140091677-140092911      4          9606   \n",
       "81675  chrY:-:1391992-1392341;1395129-1395548;1396371...      4          9606   \n",
       "81676  chrY:-:1391992-1392341;1395129-1395548;1396371...      4          9606   \n",
       "81677  chrY:-:1391992-1392341;1395129-1395548;1396371...      4          9606   \n",
       "81678  chrY:-:1391992-1392341;1395129-1395548;1396371...      4          9606   \n",
       "\n",
       "        Target Gene ID Target Gene Symbol  Target Transcript  \\\n",
       "0      ENSG00000000003             TSPAN6  ENST00000373020.9   \n",
       "1      ENSG00000000003             TSPAN6  ENST00000373020.9   \n",
       "2      ENSG00000000003             TSPAN6  ENST00000373020.9   \n",
       "3      ENSG00000000003             TSPAN6  ENST00000373020.9   \n",
       "4      ENSG00000000005               TNMD  ENST00000373031.5   \n",
       "...                ...                ...                ...   \n",
       "81674              NaN                NaN                NaN   \n",
       "81675              NaN                NaN                NaN   \n",
       "81676              NaN                NaN                NaN   \n",
       "81677              NaN                NaN                NaN   \n",
       "81678              NaN                NaN                NaN   \n",
       "\n",
       "                                 Target Reference Coords  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "81674                 NC_000023.11:-:140091677-140092911   \n",
       "81675  NC_000024.10:-:1391992-1392341;1395129-1395548...   \n",
       "81676  NC_000024.10:-:1391992-1392341;1395129-1395548...   \n",
       "81677  NC_000024.10:-:1391992-1392341;1395129-1395548...   \n",
       "81678  NC_000024.10:-:1391992-1392341;1395129-1395548...   \n",
       "\n",
       "                                            Target Alias CRISPR Mechanism  \\\n",
       "0                                                    NaN         CRISPRko   \n",
       "1                                                    NaN         CRISPRko   \n",
       "2                                                    NaN         CRISPRko   \n",
       "3                                                    NaN         CRISPRko   \n",
       "4                                                    NaN         CRISPRko   \n",
       "...                                                  ...              ...   \n",
       "81674                         chrX:-:140091677-140092911         CRISPRko   \n",
       "81675  chrY:-:1391992-1392341;1395129-1395548;1396371...         CRISPRko   \n",
       "81676  chrY:-:1391992-1392341;1395129-1395548;1396371...         CRISPRko   \n",
       "81677  chrY:-:1391992-1392341;1395129-1395548;1396371...         CRISPRko   \n",
       "81678  chrY:-:1391992-1392341;1395129-1395548;1396371...         CRISPRko   \n",
       "\n",
       "      Target Domain  ... On-Target Efficacy Score On-Target Rank  \\\n",
       "0               CDS  ...                   1.0210              2   \n",
       "1               CDS  ...                   0.9891              3   \n",
       "2               CDS  ...                   0.9622              4   \n",
       "3               CDS  ...                   0.6634              9   \n",
       "4               CDS  ...                   1.1120              1   \n",
       "...             ...  ...                      ...            ...   \n",
       "81674           CDS  ...                   0.4118             31   \n",
       "81675           CDS  ...                   0.9535              1   \n",
       "81676           CDS  ...                   0.8926              2   \n",
       "81677           CDS  ...                   0.8820              3   \n",
       "81678           CDS  ...                   0.7691              6   \n",
       "\n",
       "      Preselected As  Matching Active Arrayed Oligos  \\\n",
       "0                NaN                             NaN   \n",
       "1                NaN                             NaN   \n",
       "2                NaN                             NaN   \n",
       "3                NaN                             NaN   \n",
       "4                NaN                             NaN   \n",
       "...              ...                             ...   \n",
       "81674            NaN                             NaN   \n",
       "81675            NaN                             NaN   \n",
       "81676            NaN                             NaN   \n",
       "81677            NaN                             NaN   \n",
       "81678            NaN                             NaN   \n",
       "\n",
       "       Matching Arrayed Constructs  Pools Containing Matching Construct  \\\n",
       "0                              NaN                       CP1889, CP2115   \n",
       "1                   BRDN0003487346               CP1889, CP2114, CP2115   \n",
       "2                              NaN               CP1889, CP2114, CP2115   \n",
       "3                              NaN                                  NaN   \n",
       "4                              NaN                                  NaN   \n",
       "...                            ...                                  ...   \n",
       "81674                          NaN                                  NaN   \n",
       "81675                          NaN                                  NaN   \n",
       "81676                          NaN                                  NaN   \n",
       "81677                          NaN                                  NaN   \n",
       "81678                          NaN                                  NaN   \n",
       "\n",
       "       Pick Order Picking Round  \\\n",
       "0               1             1   \n",
       "1               2             1   \n",
       "2               3             1   \n",
       "3               4             1   \n",
       "4               1             1   \n",
       "...           ...           ...   \n",
       "81674           4             1   \n",
       "81675           1             4   \n",
       "81676           2             4   \n",
       "81677           3             4   \n",
       "81678           4             4   \n",
       "\n",
       "                                           Picking Notes   Source  \n",
       "0                                                    NaN  GENCODE  \n",
       "1                                                    NaN  GENCODE  \n",
       "2                                                    NaN  GENCODE  \n",
       "3                                                    NaN  GENCODE  \n",
       "4                                                    NaN  GENCODE  \n",
       "...                                                  ...      ...  \n",
       "81674                                                NaN    CHESS  \n",
       "81675  Previously skipped due to: Off-target CFD100 m...    CHESS  \n",
       "81676  Previously skipped due to: Off-target CFD100 m...    CHESS  \n",
       "81677  Previously skipped due to: Off-target CFD100 m...    CHESS  \n",
       "81678  Previously skipped due to: Off-target CFD100 m...    CHESS  \n",
       "\n",
       "[81679 rows x 43 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CRISPick output of three design runs: \n",
    "## 1. all protein coding genes in ensembl\n",
    "## 2. those in NCBI but not ensembl\n",
    "## 3. those unique to CHESS\n",
    "ensembl_designs=pd.read_table(\"../Data/ensembl_113_protein_coding_aggcfd_targetlocal-sgrna-designs.txt\")\n",
    "NCBI_designs=pd.read_table(\"../Data/ncbi_IDS_forjacquere-sgrna-designs.txt\")\n",
    "CHESS_designs= pd.read_table(\"../Data/chess_targetlocal_aggcfd-sgrna-designs.txt\")\n",
    "\n",
    "#replace exon ranges with identifiers for CHESS genes \n",
    "#CHESS_identifier= pd.read_csv(\"../Jacquere_library/design_annotations/CHESS3.1.3_unique_protein_coding_genes_coordinates_with_identifier.csv\",header=None)\n",
    "#CHESS_designs=CHESS_designs.merge(CHESS_identifier,left_on=\"Input\",right_on=1)\n",
    "#CHESS_designs[\"Input\"]=CHESS_designs[0]\n",
    "#CHESS_designs=CHESS_designs.drop([0,1],axis=1)\n",
    "\n",
    "jacquere_og=pd.concat([ensembl_designs,NCBI_designs,CHESS_designs],keys=[\"GENCODE\",\"RefSeq\",\"CHESS\"]).reset_index()\n",
    "jacquere_og[\"Source\"]=jacquere_og[\"level_0\"]\n",
    "jacquere_og=jacquere_og.drop([\"level_0\",\"level_1\"],axis=1)\n",
    "\n",
    "\n",
    "#remove any guides with Aggregate CFD > 4.8\n",
    "jacquere=jacquere_og[jacquere_og[\"Aggregate CFD Score\"]!=\"MAX\"].reset_index(drop=True)\n",
    "jacquere[\"Aggregate CFD Score\"]=jacquere[\"Aggregate CFD Score\"].astype(float)\n",
    "jacquere=jacquere[jacquere[\"Aggregate CFD Score\"]<=4.8].reset_index(drop=True)\n",
    "\n",
    "\n",
    "jacquere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "51e41e20-5185-4992-b6c8-3671f6af9007",
   "metadata": {},
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "[Errno 89] Operation canceled",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mOSError\u001b[39m                                   Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[3]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m#negative controls: 900 intergenic and 100 nontargeting\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m intergenics=\u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m../Data/Negative Controls/jacquere_900_intergenics.csv\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      4\u001b[39m intergenics[\u001b[33m\"\u001b[39m\u001b[33mInput\u001b[39m\u001b[33m\"\u001b[39m]=\u001b[33m\"\u001b[39m\u001b[33mONE_SITE_INTERGENIC\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      5\u001b[39m intergenics[\u001b[33m\"\u001b[39m\u001b[33mSource\u001b[39m\u001b[33m\"\u001b[39m]=\u001b[33m\"\u001b[39m\u001b[33mONE_SITE_INTERGENIC\u001b[39m\u001b[33m\"\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Library/CloudStorage/GoogleDrive-ldrepano@broadinstitute.org/Shared drives/GPP Cloud /R&D/People/Laura/CRISPick Parameter Tuning/GPP-Jacquere/run_jacquere_code/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1026\u001b[39m, in \u001b[36mread_csv\u001b[39m\u001b[34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[39m\n\u001b[32m   1013\u001b[39m kwds_defaults = _refine_defaults_read(\n\u001b[32m   1014\u001b[39m     dialect,\n\u001b[32m   1015\u001b[39m     delimiter,\n\u001b[32m   (...)\u001b[39m\u001b[32m   1022\u001b[39m     dtype_backend=dtype_backend,\n\u001b[32m   1023\u001b[39m )\n\u001b[32m   1024\u001b[39m kwds.update(kwds_defaults)\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Library/CloudStorage/GoogleDrive-ldrepano@broadinstitute.org/Shared drives/GPP Cloud /R&D/People/Laura/CRISPick Parameter Tuning/GPP-Jacquere/run_jacquere_code/lib/python3.13/site-packages/pandas/io/parsers/readers.py:620\u001b[39m, in \u001b[36m_read\u001b[39m\u001b[34m(filepath_or_buffer, kwds)\u001b[39m\n\u001b[32m    617\u001b[39m _validate_names(kwds.get(\u001b[33m\"\u001b[39m\u001b[33mnames\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[32m    619\u001b[39m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m620\u001b[39m parser = \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    622\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[32m    623\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Library/CloudStorage/GoogleDrive-ldrepano@broadinstitute.org/Shared drives/GPP Cloud /R&D/People/Laura/CRISPick Parameter Tuning/GPP-Jacquere/run_jacquere_code/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1620\u001b[39m, in \u001b[36mTextFileReader.__init__\u001b[39m\u001b[34m(self, f, engine, **kwds)\u001b[39m\n\u001b[32m   1617\u001b[39m     \u001b[38;5;28mself\u001b[39m.options[\u001b[33m\"\u001b[39m\u001b[33mhas_index_names\u001b[39m\u001b[33m\"\u001b[39m] = kwds[\u001b[33m\"\u001b[39m\u001b[33mhas_index_names\u001b[39m\u001b[33m\"\u001b[39m]\n\u001b[32m   1619\u001b[39m \u001b[38;5;28mself\u001b[39m.handles: IOHandles | \u001b[38;5;28;01mNone\u001b[39;00m = \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m1620\u001b[39m \u001b[38;5;28mself\u001b[39m._engine = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Library/CloudStorage/GoogleDrive-ldrepano@broadinstitute.org/Shared drives/GPP Cloud /R&D/People/Laura/CRISPick Parameter Tuning/GPP-Jacquere/run_jacquere_code/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1898\u001b[39m, in \u001b[36mTextFileReader._make_engine\u001b[39m\u001b[34m(self, f, engine)\u001b[39m\n\u001b[32m   1895\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[32m   1897\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m1898\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmapping\u001b[49m\u001b[43m[\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m]\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1899\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[32m   1900\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.handles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Library/CloudStorage/GoogleDrive-ldrepano@broadinstitute.org/Shared drives/GPP Cloud /R&D/People/Laura/CRISPick Parameter Tuning/GPP-Jacquere/run_jacquere_code/lib/python3.13/site-packages/pandas/io/parsers/c_parser_wrapper.py:93\u001b[39m, in \u001b[36mCParserWrapper.__init__\u001b[39m\u001b[34m(self, src, **kwds)\u001b[39m\n\u001b[32m     90\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m kwds[\u001b[33m\"\u001b[39m\u001b[33mdtype_backend\u001b[39m\u001b[33m\"\u001b[39m] == \u001b[33m\"\u001b[39m\u001b[33mpyarrow\u001b[39m\u001b[33m\"\u001b[39m:\n\u001b[32m     91\u001b[39m     \u001b[38;5;66;03m# Fail here loudly instead of in cython after reading\u001b[39;00m\n\u001b[32m     92\u001b[39m     import_optional_dependency(\u001b[33m\"\u001b[39m\u001b[33mpyarrow\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m93\u001b[39m \u001b[38;5;28mself\u001b[39m._reader = \u001b[43mparsers\u001b[49m\u001b[43m.\u001b[49m\u001b[43mTextReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43msrc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     95\u001b[39m \u001b[38;5;28mself\u001b[39m.unnamed_cols = \u001b[38;5;28mself\u001b[39m._reader.unnamed_cols\n\u001b[32m     97\u001b[39m \u001b[38;5;66;03m# error: Cannot determine type of 'names'\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:574\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader.__cinit__\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:663\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._get_header\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:874\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._tokenize_rows\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:891\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._check_tokenize_status\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:2053\u001b[39m, in \u001b[36mpandas._libs.parsers.raise_parser_error\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[31mOSError\u001b[39m: [Errno 89] Operation canceled"
     ]
    }
   ],
   "source": [
    "#negative controls: 900 intergenic and 100 nontargeting\n",
    "intergenics=pd.read_csv(\"../Data/Negative Controls/jacquere_900_intergenics.csv\")\n",
    "\n",
    "intergenics[\"Input\"]=\"ONE_SITE_INTERGENIC\"\n",
    "intergenics[\"Source\"]=\"ONE_SITE_INTERGENIC\"\n",
    "intergenics[\"Target Gene Symbol\"]=\"ONE_SITE_INTERGENIC\"\n",
    "intergenics[\"On-Target Ruleset\"]=\"RS3seq-Chen2013+RS3target\"\n",
    "intergenics[\"Target Gene ID\"]=\"ONE_SITE_INTERGENIC\"\n",
    "\n",
    "nontargeting= pd.read_table(\"../Data/Negative Controls/jacquere_100_nontargeting-sgrna-designs.txt\")\n",
    "nontargeting=nontargeting[nontargeting[\"Target Gene Symbol\"].isna()].reset_index(drop=True) #remove the one targeting guide required for CRISPick to run \n",
    "nontargeting[\"Input\"]=\"NO_SITE\"\n",
    "nontargeting[\"Source\"]=\"NO_SITE\"\n",
    "nontargeting[\"Target Gene Symbol\"]=\"NO_SITE\"\n",
    "nontargeting[\"Target Gene ID\"]=\"NO_SITE\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9876fcb-411e-4caf-a23f-3f2b657f362f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#exporting jacquere prior to reducing the quota to support screening of jacquere with variable # guides per gene\n",
    "jacquere_quotasubsampling=jacquere[[\"Target Gene Symbol\",\"Input\",\"Source\",\"Pick Order\",\"sgRNA Sequence\"]].copy()\n",
    "jacquere_quotasubsampling.loc[jacquere[\"Target Gene Symbol\"].isna(),\"Target Gene Symbol\"]=\"N/A\"\n",
    "\n",
    "intergenics[\"Pick Order\"]=\"N/A\"\n",
    "nontargeting[\"Pick Order\"]=\"N/A\"\n",
    "jacquere_quotasubsampling=pd.concat([jacquere_quotasubsampling,intergenics[[\"Target Gene Symbol\",\"Input\",\"Source\",\"Pick Order\",\"sgRNA Sequence\"]],nontargeting[[\"Target Gene Symbol\",\"Input\",\"Source\",\"Pick Order\",\"sgRNA Sequence\"]]])\n",
    "#jacquere_quotasubsampling.to_csv(\"../Jacquere_uptoquota4.csv\",index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a84853ba-1da3-40e3-b740-bdbd0b50e0a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "jacquere[\"On-Target Efficacy Score\"].hist(bins=50,grid=False)\n",
    "plt.axvline(x=0.2,c=\"red\",linestyle=\"--\",label=\"RS3=0.2\")\n",
    "plt.title(\"distribution of rs3 scores in jacquere (4 guides/gene)\")\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "127cc833-9966-44bb-9d96-1c30537db464",
   "metadata": {},
   "source": [
    "[Picking rounds](https://portals.broadinstitute.org/gppx/crispick/platform/how-it-works). Round 1 or 2 is ideal (criteria relaxes from 5-65 to 5-80 along the protein coding region, both are acceptable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "781c8a5c-14ac-4367-9f69-288575fb210a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"% of Jacquere in picking round 1 or 2:\", 100*len(jacquere[jacquere[\"Picking Round\"].isin([1,2])])/len(jacquere))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf8cc7af-144b-4697-8dc8-b889dd879c17",
   "metadata": {},
   "source": [
    "There are distinct Ensembl IDs that associated with the same Gene Symbol, but are optimally targeted by distinct sets of guides. These should be kept separate in Jacquere, and thus our suggested mappings should use Ensembl IDs. Note that the values below represent the 4 guide/gene library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "415d8c69-035a-463f-9e19-0a0aed087944",
   "metadata": {},
   "outputs": [],
   "source": [
    "jacquere_ensembl=jacquere[jacquere[\"Source\"]==\"GENCODE\"]\n",
    "gene_symbol_counts=pd.DataFrame(jacquere_ensembl[\"Target Gene Symbol\"].value_counts())\n",
    "same_genesymbol_diff_ID_symbols=gene_symbol_counts[gene_symbol_counts[\"count\"]>4].reset_index()[\"Target Gene Symbol\"]\n",
    "print(\"# gene symbols associated with >1 ensembl gene ID\", len(same_genesymbol_diff_ID_symbols))\n",
    "#get all sgRNAs (up to 4) associated with a given target\n",
    "guides_per_input=pd.DataFrame(jacquere.groupby(['Input','Target Gene Symbol','Source'])['sgRNA Sequence'].apply(list)).reset_index()\n",
    "unique_guidesets_shared_genesymbol=pd.DataFrame(guides_per_input[guides_per_input[\"Target Gene Symbol\"].isin(same_genesymbol_diff_ID_symbols)][\"sgRNA Sequence\"].value_counts())\n",
    "print(\"% of genes that share a gene symbol that have unique sets of guides\", 100*len(unique_guidesets_shared_genesymbol[unique_guidesets_shared_genesymbol[\"count\"]==1])/len(unique_guidesets_shared_genesymbol))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6296c9a-9eb0-43b7-a162-05753008b646",
   "metadata": {},
   "source": [
    "## Determining Ideal Quota for Jacquere\n",
    "\n",
    "Subsampling analysis, identifying how top 1 vs 2 vs 3 vs 4 guides in Jacquere are able to yield gene z-scores appropriate for their predicted viability effect (essential vs nonessential) in Deweirdt 2022 tiling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cefcde61-08f0-4c19-8154-0b4d33afbca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "rs3val_zscores= pd.read_csv(\"../../2. Developing promiscuous classifier/Data/rs3valdata_updatedmappings_zscore.csv\")\n",
    "rs3val_zscores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81a59d7e-a1cb-44c8-9e34-eda462987447",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get essential vs nonessential classifications \n",
    "\n",
    "noness_genes = pd.read_csv(\"../../2. Developing promiscuous classifier/Data/AchillesNonessentialControls.csv\")\n",
    "noness_genes[\"Gene\"]=noness_genes[\"Gene\"].apply(lambda x: x.split(\"(\")[0])\n",
    "#remove space from gene name \n",
    "noness_genes[\"Gene\"]=noness_genes[\"Gene\"].str.replace(\" \",\"\")\n",
    "noness_gene_list=noness_genes[\"Gene\"].tolist()\n",
    "\n",
    "essential_genes = pd.read_csv(\"../../2. Developing promiscuous classifier/Data/AchillesCommonEssentialControls.csv\")\n",
    "essential_genes[\"Gene\"]=essential_genes[\"Gene\"].apply(lambda x: x.split(\"(\")[0])\n",
    "#remove space from gene name \n",
    "essential_genes[\"Gene\"]=essential_genes[\"Gene\"].str.replace(\" \",\"\")\n",
    "essential_gene_list=essential_genes[\"Gene\"].tolist()\n",
    "\n",
    "rs3val_zscores[\"is_essential\"]=rs3val_zscores[\"On-target Gene Symbols\"].isin(essential_gene_list)\n",
    "rs3val_zscores[\"is_nonessential\"]=rs3val_zscores[\"On-target Gene Symbols\"].isin(noness_gene_list)\n",
    "\n",
    "print(\"# essential:\",len(rs3val_zscores[rs3val_zscores[\"is_essential\"]]))\n",
    "print(\"# nonessential:\",len(rs3val_zscores[rs3val_zscores[\"is_nonessential\"]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50c233c8-24ae-4b39-83e5-b84b590409c5",
   "metadata": {},
   "source": [
    "Get library subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a02b167-0592-4471-b137-5c3df638bb05",
   "metadata": {},
   "outputs": [],
   "source": [
    "jacquere_with_z_scores=pd.merge(left=jacquere,right=rs3val_zscores,on=\"sgRNA Sequence\")\n",
    "\n",
    "jacquere_top1=jacquere_with_z_scores[jacquere_with_z_scores[\"Pick Order\"]==1].reset_index(drop=True)\n",
    "jacquere_top2=jacquere_with_z_scores[jacquere_with_z_scores[\"Pick Order\"].isin([1,2])].reset_index(drop=True)\n",
    "jacquere_top3=jacquere_with_z_scores[jacquere_with_z_scores[\"Pick Order\"].isin([1,2,3])].reset_index(drop=True)\n",
    "jacquere_top4=jacquere_with_z_scores\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a74127a3-5aa9-4887-ba63-0f6c8a829331",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get average z-score per gene, separately for 1 vs 2 vs 3 vs 4 guides per gene in Jacquere\n",
    "# not using stouffer's b/c biased for gene set size \n",
    "\n",
    "gene_avg_z_jacqueretop1 = (jacquere_top1.groupby(['Target Gene Symbol',\"is_essential\",\"is_nonessential\"])\n",
    "                     .agg(mean_z = ('z_scored_sgRNA_lfc', 'mean'))\n",
    "                     .reset_index())\n",
    "\n",
    "gene_avg_z_jacqueretop2 = (jacquere_top2.groupby(['Target Gene Symbol',\"is_essential\",\"is_nonessential\"])\n",
    "                     .agg(mean_z = ('z_scored_sgRNA_lfc', 'mean'),\n",
    "                          n_sgrnas = ('z_scored_sgRNA_lfc', 'count'))\n",
    "                     .reset_index())\n",
    "\n",
    "gene_avg_z_jacqueretop3 = (jacquere_top3.groupby(['Target Gene Symbol',\"is_essential\",\"is_nonessential\"])\n",
    "                     .agg(mean_z = ('z_scored_sgRNA_lfc', 'mean'),\n",
    "                          n_sgrnas = ('z_scored_sgRNA_lfc', 'count'))\n",
    "                     .reset_index())\n",
    "\n",
    "gene_avg_z_jacqueretop4 = (jacquere_top4.groupby(['Target Gene Symbol',\"is_essential\",\"is_nonessential\"])\n",
    "                     .agg(mean_z = ('z_scored_sgRNA_lfc', 'mean'),\n",
    "                          n_sgrnas = ('z_scored_sgRNA_lfc', 'count'))\n",
    "                     .reset_index())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1447102-037d-4e83-a8dc-0edf8353e6ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add randomly picked guides to comparison\n",
    "random.seed(1)\n",
    "sgRNAs_selected=[]\n",
    "for gene in rs3val_zscores[\"On-target Gene Symbols\"].unique():\n",
    "    #all sgRNAs that target that gene \n",
    "    gene_sgRNAs=rs3val_zscores[rs3val_zscores[\"On-target Gene Symbols\"]==gene][\"sgRNA Sequence\"].tolist()\n",
    "    #select 4 of these sgRNAs randomly\n",
    "    if len(gene_sgRNAs)>=4:  \n",
    "        random_4=random.sample(gene_sgRNAs,4)\n",
    "    sgRNAs_selected=sgRNAs_selected+random_4\n",
    "random_library=rs3val_zscores[rs3val_zscores[\"sgRNA Sequence\"].isin(sgRNAs_selected)].reset_index(drop=True).copy()\n",
    "\n",
    "gene_avg_z_random_4 = (random_library.groupby(['On-target Gene Symbols',\"is_essential\",\"is_nonessential\"])\n",
    "                     .agg(mean_z = ('z_scored_sgRNA_lfc', 'mean'),\n",
    "                          n_sgrnas = ('z_scored_sgRNA_lfc', 'count'))\n",
    "                     .reset_index())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db527539-d78f-44b7-9f67-0e6cbee0a7a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add Brunello to comparison\n",
    "brunello= pd.read_excel(\"../../5. Assessment of  CRISPRko Cas9 Genome Wide Libraries/Data/Original Annotations of CRISPRko Cas9 Genome Wide Libraries/STable 21 Brunello.xlsx\",sheet_name=\"Sheet1\")\n",
    "brunello_guides=brunello[\"sgRNA Target Sequence\"].tolist()\n",
    "brunello_with_z_scores=rs3val_zscores[rs3val_zscores[\"sgRNA Sequence\"].isin(brunello_guides)].reset_index(drop=True)\n",
    "\n",
    "\n",
    "gene_avg_z_brunello = (brunello_with_z_scores.groupby(['On-target Gene Symbols',\"is_essential\",\"is_nonessential\"])\n",
    "                     .agg(mean_z = ('z_scored_sgRNA_lfc', 'mean'),\n",
    "                          n_sgrnas = ('z_scored_sgRNA_lfc', 'count'))\n",
    "                     .reset_index())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "960e0f9f-4755-4129-8721-3af194558ec3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add Gattinara to comparison\n",
    "gattinara= pd.read_table(\"../../5. Assessment of  CRISPRko Cas9 Genome Wide Libraries/Data/Original Annotations of CRISPRko Cas9 Genome Wide Libraries/broadgpp-gattinara-library-contents.txt\")\n",
    "gattinara_guides=gattinara[gattinara[\"Annotated Gene Symbol\"].str.count(\"NEG_CONTROL\")==0][\"Barcode Sequence\"].tolist()\n",
    "gattinara_with_z_scores=rs3val_zscores[rs3val_zscores[\"sgRNA Sequence\"].isin(gattinara_guides)].reset_index(drop=True)\n",
    "gene_avg_z_gattinara = (gattinara_with_z_scores.groupby(['On-target Gene Symbols',\"is_essential\",\"is_nonessential\"])\n",
    "                     .agg(mean_z = ('z_scored_sgRNA_lfc', 'mean'),\n",
    "                          n_sgrnas = ('z_scored_sgRNA_lfc', 'count'))\n",
    "                     .reset_index())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fce0a61-3dd5-47cc-8d0d-e2c4277158de",
   "metadata": {},
   "source": [
    "Get false negative (% of essential genes with z-score not passing threshold) rates to assess optimal quota"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c148dee-6243-4e0d-af19-75ee52152d71",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_avg_z_jacqueretop1_essentials=gene_avg_z_jacqueretop1[gene_avg_z_jacqueretop1[\"is_essential\"]].reset_index(drop=True)\n",
    "gene_avg_z_jacqueretop1_essentials[\"z<-2\"]=gene_avg_z_jacqueretop1_essentials[\"mean_z\"]<-2\n",
    "gene_avg_z_jacqueretop1_essentials[\"z<-3\"]=gene_avg_z_jacqueretop1_essentials[\"mean_z\"]<-3\n",
    "gene_avg_z_jacqueretop1_essentials[\"z<-4\"]=gene_avg_z_jacqueretop1_essentials[\"mean_z\"]<-4\n",
    "jacqueretop1_FNR_zneg2=len(gene_avg_z_jacqueretop1_essentials[gene_avg_z_jacqueretop1_essentials[\"z<-2\"]==False])/len(gene_avg_z_jacqueretop1_essentials)\n",
    "jacqueretop1_FNR_zneg3=len(gene_avg_z_jacqueretop1_essentials[gene_avg_z_jacqueretop1_essentials[\"z<-3\"]==False])/len(gene_avg_z_jacqueretop1_essentials)\n",
    "jacqueretop1_FNR_zneg4=len(gene_avg_z_jacqueretop1_essentials[gene_avg_z_jacqueretop1_essentials[\"z<-4\"]==False])/len(gene_avg_z_jacqueretop1_essentials)\n",
    "\n",
    "gene_avg_z_jacqueretop2_essentials=gene_avg_z_jacqueretop2[gene_avg_z_jacqueretop2[\"is_essential\"]].reset_index(drop=True)\n",
    "gene_avg_z_jacqueretop2_essentials[\"z<-2\"]=gene_avg_z_jacqueretop2_essentials[\"mean_z\"]<-2\n",
    "gene_avg_z_jacqueretop2_essentials[\"z<-3\"]=gene_avg_z_jacqueretop2_essentials[\"mean_z\"]<-3\n",
    "gene_avg_z_jacqueretop2_essentials[\"z<-4\"]=gene_avg_z_jacqueretop2_essentials[\"mean_z\"]<-4\n",
    "jacqueretop2_FNR_zneg2=len(gene_avg_z_jacqueretop2_essentials[gene_avg_z_jacqueretop2_essentials[\"z<-2\"]==False])/len(gene_avg_z_jacqueretop2_essentials)\n",
    "jacqueretop2_FNR_zneg3=len(gene_avg_z_jacqueretop2_essentials[gene_avg_z_jacqueretop2_essentials[\"z<-3\"]==False])/len(gene_avg_z_jacqueretop2_essentials)\n",
    "jacqueretop2_FNR_zneg4=len(gene_avg_z_jacqueretop2_essentials[gene_avg_z_jacqueretop2_essentials[\"z<-4\"]==False])/len(gene_avg_z_jacqueretop2_essentials)\n",
    "\n",
    "gene_avg_z_jacqueretop3_essentials=gene_avg_z_jacqueretop3[gene_avg_z_jacqueretop3[\"is_essential\"]].reset_index(drop=True)\n",
    "gene_avg_z_jacqueretop3_essentials[\"z<-2\"]=gene_avg_z_jacqueretop3_essentials[\"mean_z\"]<-2\n",
    "gene_avg_z_jacqueretop3_essentials[\"z<-3\"]=gene_avg_z_jacqueretop3_essentials[\"mean_z\"]<-3\n",
    "gene_avg_z_jacqueretop3_essentials[\"z<-4\"]=gene_avg_z_jacqueretop3_essentials[\"mean_z\"]<-4\n",
    "jacqueretop3_FNR_zneg2=len(gene_avg_z_jacqueretop3_essentials[gene_avg_z_jacqueretop3_essentials[\"z<-2\"]==False])/len(gene_avg_z_jacqueretop3_essentials)\n",
    "jacqueretop3_FNR_zneg3=len(gene_avg_z_jacqueretop3_essentials[gene_avg_z_jacqueretop3_essentials[\"z<-3\"]==False])/len(gene_avg_z_jacqueretop3_essentials)\n",
    "jacqueretop3_FNR_zneg4=len(gene_avg_z_jacqueretop3_essentials[gene_avg_z_jacqueretop3_essentials[\"z<-4\"]==False])/len(gene_avg_z_jacqueretop3_essentials)\n",
    "\n",
    "gene_avg_z_jacqueretop4_essentials=gene_avg_z_jacqueretop4[gene_avg_z_jacqueretop4[\"is_essential\"]].reset_index(drop=True)\n",
    "gene_avg_z_jacqueretop4_essentials[\"z<-2\"]=gene_avg_z_jacqueretop4_essentials[\"mean_z\"]<-2\n",
    "gene_avg_z_jacqueretop4_essentials[\"z<-3\"]=gene_avg_z_jacqueretop4_essentials[\"mean_z\"]<-3\n",
    "gene_avg_z_jacqueretop4_essentials[\"z<-4\"]=gene_avg_z_jacqueretop4_essentials[\"mean_z\"]<-4\n",
    "jacqueretop4_FNR_zneg2=len(gene_avg_z_jacqueretop4_essentials[gene_avg_z_jacqueretop4_essentials[\"z<-2\"]==False])/len(gene_avg_z_jacqueretop4_essentials)\n",
    "jacqueretop4_FNR_zneg3=len(gene_avg_z_jacqueretop4_essentials[gene_avg_z_jacqueretop4_essentials[\"z<-3\"]==False])/len(gene_avg_z_jacqueretop4_essentials)\n",
    "jacqueretop4_FNR_zneg4=len(gene_avg_z_jacqueretop4_essentials[gene_avg_z_jacqueretop4_essentials[\"z<-4\"]==False])/len(gene_avg_z_jacqueretop4_essentials)\n",
    "\n",
    "gene_avg_z_random_4_essentials=gene_avg_z_random_4[gene_avg_z_random_4[\"is_essential\"]].reset_index(drop=True)\n",
    "gene_avg_z_random_4_essentials[\"z<-2\"]=gene_avg_z_random_4_essentials[\"mean_z\"]<-2\n",
    "gene_avg_z_random_4_essentials[\"z<-3\"]=gene_avg_z_random_4_essentials[\"mean_z\"]<-3\n",
    "gene_avg_z_random_4_essentials[\"z<-4\"]=gene_avg_z_random_4_essentials[\"mean_z\"]<-4\n",
    "random_4_FNR_zneg2=len(gene_avg_z_random_4_essentials[gene_avg_z_random_4_essentials[\"z<-2\"]==False])/len(gene_avg_z_random_4_essentials)\n",
    "random_4_FNR_zneg3=len(gene_avg_z_random_4_essentials[gene_avg_z_random_4_essentials[\"z<-3\"]==False])/len(gene_avg_z_random_4_essentials)\n",
    "random_4_FNR_zneg4=len(gene_avg_z_random_4_essentials[gene_avg_z_random_4_essentials[\"z<-4\"]==False])/len(gene_avg_z_random_4_essentials)\n",
    "\n",
    "gene_avg_z_brunello_essentials=gene_avg_z_brunello[gene_avg_z_brunello[\"is_essential\"]].reset_index(drop=True)\n",
    "gene_avg_z_brunello_essentials[\"z<-2\"]=gene_avg_z_brunello_essentials[\"mean_z\"]<-2\n",
    "gene_avg_z_brunello_essentials[\"z<-3\"]=gene_avg_z_brunello_essentials[\"mean_z\"]<-3\n",
    "gene_avg_z_brunello_essentials[\"z<-4\"]=gene_avg_z_brunello_essentials[\"mean_z\"]<-4\n",
    "brunello_FNR_zneg2=len(gene_avg_z_brunello_essentials[gene_avg_z_brunello_essentials[\"z<-2\"]==False])/len(gene_avg_z_brunello_essentials)\n",
    "brunello_FNR_zneg3=len(gene_avg_z_brunello_essentials[gene_avg_z_brunello_essentials[\"z<-3\"]==False])/len(gene_avg_z_brunello_essentials)\n",
    "brunello_FNR_zneg4=len(gene_avg_z_brunello_essentials[gene_avg_z_brunello_essentials[\"z<-4\"]==False])/len(gene_avg_z_brunello_essentials)\n",
    "\n",
    "\n",
    "gene_avg_z_gattinara_essentials=gene_avg_z_gattinara[gene_avg_z_gattinara[\"is_essential\"]].reset_index(drop=True)\n",
    "gene_avg_z_gattinara_essentials[\"z<-2\"]=gene_avg_z_gattinara_essentials[\"mean_z\"]<-2\n",
    "gene_avg_z_gattinara_essentials[\"z<-3\"]=gene_avg_z_gattinara_essentials[\"mean_z\"]<-3\n",
    "gene_avg_z_gattinara_essentials[\"z<-4\"]=gene_avg_z_gattinara_essentials[\"mean_z\"]<-4\n",
    "gattinara_FNR_zneg2=len(gene_avg_z_gattinara_essentials[gene_avg_z_gattinara_essentials[\"z<-2\"]==False])/len(gene_avg_z_gattinara_essentials)\n",
    "gattinara_FNR_zneg3=len(gene_avg_z_gattinara_essentials[gene_avg_z_gattinara_essentials[\"z<-3\"]==False])/len(gene_avg_z_gattinara_essentials)\n",
    "gattinara_FNR_zneg4=len(gene_avg_z_gattinara_essentials[gene_avg_z_gattinara_essentials[\"z<-4\"]==False])/len(gene_avg_z_gattinara_essentials)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd413e03-fcf1-4278-82f3-d79330c14e19",
   "metadata": {},
   "outputs": [],
   "source": [
    "libraries=[\"Random \\nQuota=4\",\"Brunello \\nQuota=4\",\"Gattinara \\nQuota=2\",\"Jacquere \\nQuota=1\",\"Jacquere \\nQuota=2\",\"Jacquere \\nQuota=3\",\"Jacquere \\nQuota=4\"]\n",
    "FNR_df=pd.DataFrame({\"Z<-2\":[random_4_FNR_zneg2,brunello_FNR_zneg2,gattinara_FNR_zneg2,jacqueretop1_FNR_zneg2,jacqueretop2_FNR_zneg2,jacqueretop3_FNR_zneg2,jacqueretop4_FNR_zneg2],\n",
    "             \"Z<-3\":[random_4_FNR_zneg3,brunello_FNR_zneg3,gattinara_FNR_zneg3,jacqueretop1_FNR_zneg3,jacqueretop2_FNR_zneg3,jacqueretop3_FNR_zneg3,jacqueretop4_FNR_zneg3],\n",
    "              \"Z<-4\":[random_4_FNR_zneg4,brunello_FNR_zneg4,gattinara_FNR_zneg4,jacqueretop1_FNR_zneg4,jacqueretop2_FNR_zneg4,jacqueretop3_FNR_zneg4,jacqueretop4_FNR_zneg4]\n",
    "             },index=libraries)\n",
    "FNR_df=FNR_df.multiply(100).astype(int) #convert decimals to percentages\n",
    "FNR_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aef050e5-3d8b-4ee4-b85d-65888394f793",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = np.arange(len(libraries))  # the label locations\n",
    "width = 0.25  # the width of the bars\n",
    "multiplier = 0\n",
    "\n",
    "fig, ax = plt.subplots(layout='constrained')\n",
    "ax.set_prop_cycle(cycler('color', plt.cm.tab20.colors))\n",
    "\n",
    "for zcutoff, FNR in FNR_df.items():\n",
    "    offset = width * multiplier\n",
    "    rects = ax.bar(x + offset, FNR, width, label=zcutoff)\n",
    "    ax.bar_label(rects, padding=3)\n",
    "    multiplier += 1\n",
    "\n",
    "# Add some text for labels, title and custom x-axis tick labels, etc.\n",
    "ax.set_ylabel('False Negative Rate (%)',fontsize=16)\n",
    "ax.set_title('Identifying essential genes from mean sgRNA Z-score',fontsize=16)\n",
    "ax.set_xticks(x + width, libraries)\n",
    "ax.legend(loc='upper right', ncols=3,title=\"Cutoff to classify gene as essential\")\n",
    "ax.set_yticks([])\n",
    "ax.set_ylim([0,FNR_df.max().max()+5])\n",
    "ax.set_xlabel(\"sgRNA Library\",fontsize=16)\n",
    "#plt.savefig(\"../Figures/jacquere_subsampling_FNR.png\",bbox_inches=\"tight\",dpi=600)\n",
    "#data source: DeWeirdt 2022 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41fe23dc-f80e-4662-8247-452fb8c9923e",
   "metadata": {},
   "source": [
    "Note that this plot should **not** be interpreted as to suggest that **statistical significance** decreases with the use of more guides; in reality, screens are not interpreted with the average z-score across guides, but rather a downstream analysis method that assigns greater signficance to genes with evidence from numerous guides. Rather, this plot reveals that the inclusion of 4th guide picks decreases guide quality, thus incorporating more guides that may not necessarily reflect the true phenotype of the gene. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c25a5ba4-059e-45e0-be7e-6a9775f93f18",
   "metadata": {},
   "source": [
    "## Generate outputs of Jacquere, assembled"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd75303a-772c-43ac-ba47-44e3b605bc1f",
   "metadata": {},
   "source": [
    "Subset to the top 3 guide picks, join with the negative controls, and generate an output file showing the optimal guide picks for each target. Features repeated guides/rows since certain guides are picked to target multiple genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b51d0f0-321e-453f-9369-aef91c510b25",
   "metadata": {},
   "outputs": [],
   "source": [
    "jacquere_quota3=jacquere[jacquere[\"Pick Order\"].isin([1,2,3])].reset_index(drop=True)\n",
    "#export for comparison to other libraries\n",
    "#jacquere_quota3.to_csv(\"../Data/jacquere_assembled_crispick.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad6b7589-3372-4433-aec5-a8b185f15ea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "jacquere_export_columns=[\"Input\",\"Target Gene Symbol\",\"sgRNA Sequence\",\"On-Target Ruleset\",\"On-Target Efficacy Score\",\"Aggregate CFD Score\",\"Source\"]\n",
    "jacquere_quota3_export=jacquere_quota3[jacquere_export_columns]\n",
    "negative_controls=pd.concat([intergenics[jacquere_export_columns],nontargeting[jacquere_export_columns]])\n",
    "jacquere_quota3_export=pd.concat([jacquere_quota3_export,negative_controls])\n",
    "\n",
    "jacquere_quota3_export=jacquere_quota3_export.rename(columns={\"Input\":\"Target Gene ID\"})\n",
    "#jacquere_quota3_export.to_csv(\"../Jacquere_PerTargetAnnotations.csv\",index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f17d9dd5-64af-4aef-b649-acde6dffbd2b",
   "metadata": {},
   "source": [
    "Report as well the intended targets as **sets** of guides as to reveal which genes are intended to be targeted by the same optimal set of guides. This is an artifact of selecting the \"target local guide selection\" option in CRISPick"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df0d6c6c-baa3-4388-902d-065d97473969",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get # guides picked to target >1 gene\n",
    "guiderepeatcounts=pd.DataFrame(jacquere_quota3[\"sgRNA Sequence\"].value_counts())\n",
    "print(\"% of guides picked for exactly 1 gene:\", 100-100*len(guiderepeatcounts[guiderepeatcounts[\"count\"]>1])/len(jacquere_quota3))\n",
    "\n",
    "# Get genes that share identical sets of guides\n",
    "guides_per_input=pd.DataFrame(jacquere_quota3.groupby(['Input','Target Gene Symbol','Source'])['sgRNA Sequence'].apply(lambda x:\"_\".join(sorted(x)))).reset_index()\n",
    "guides_per_input[\"guide set\"]=guides_per_input[\"sgRNA Sequence\"]\n",
    "\n",
    "repeated_guide_sets=pd.DataFrame(guides_per_input.groupby([\"guide set\"]).agg(geneSymbols_targeted=(\"Target Gene Symbol\",list),geneIDs_targeted=(\"Input\",list),source_genes_targeted=(\"Source\",list),n_genes_targeted=(\"Target Gene Symbol\",\"count\"))).sort_values(by=\"n_genes_targeted\",ascending=False).reset_index()\n",
    "#repeated_guide_sets.to_csv(\"../Jacquere_PerGuideSetAnnotations.csv\",index=False)\n",
    "\n",
    "#report how many guides were picked for 1 vs. 2 etc genes\n",
    "guide_set_repeat_counts=pd.DataFrame(guides_per_input[\"sgRNA Sequence\"].value_counts())\n",
    "guide_set_repeat_counts[\"count\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa0b3df4-eddf-4f11-bac0-db258d348bd0",
   "metadata": {},
   "source": [
    "Generate reports detailing **all NCBI and Ensembl genes that guides map to** rather than just the genes that each guide is *intended* to target. Mappings from internal GPP LIMS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61cffac6-ca5f-457d-b839-83689d7e6bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Jacquere guides mapped back to ensembl to see if untargeted genes are targeted by guides intended for other genes\n",
    "jacquere_ensembl_mapped= pd.read_csv(\"../../5. Assessment of  CRISPRko Cas9 Genome Wide Libraries/Data/GENCODE47 Mappings of CRISPRko Cas9 Genome Wide Libraries/jacquere_quota3_adhoc_sgRNA_disco_GRCh38_Ensembl_SpyoCas9Ko_strict.csv\")\n",
    "#Above, but mapped to NCBI\n",
    "jacquere_NCBI_mapped= pd.read_csv(\"../Data/jacquere_quota3_adhoc_sgRNA_disco_GRCh38_NCBI_SpyoCas9Ko_strict.csv\")\n",
    "\n",
    "\n",
    "jacquere_quota3=jacquere_quota3.merge(jacquere_ensembl_mapped[[\"Target Sequence\",\"On-target Gene Symbols\",\"On-target Gene IDs\"]],\n",
    "                      left_on=\"sgRNA Sequence\", right_on=\"Target Sequence\",\n",
    "                      how=\"left\").drop([\"Target Sequence\"],axis=1).rename(columns={\"On-target Gene Symbols\":\"Ensembl.v113 mapped gene symbols\",\n",
    "                                                                                  \"On-target Gene IDs\":\"Ensembl.v113 mapped gene IDs\"})\n",
    "jacquere_quota3=jacquere_quota3.merge(jacquere_NCBI_mapped[[\"Target Sequence\",\"On-target Gene Symbols\",\"On-target Gene IDs\"]],\n",
    "                      left_on=\"sgRNA Sequence\", right_on=\"Target Sequence\",\n",
    "                      how=\"left\").drop([\"Target Sequence\"],axis=1).rename(columns={\"On-target Gene Symbols\":\"NCBI 08-2024 mapped gene symbols\",\n",
    "                                                                                  \"On-target Gene IDs\":\"NCBI 08-2024 mapped gene IDs\"})\n",
    "\n",
    "jacquere_quota3=jacquere_quota3.fillna(\"\")\n",
    "jacquere_quota3[\"Gene targeted\"]=np.where(jacquere_quota3[\"Target Gene Symbol\"].isna(), jacquere_quota3[\"Input\"],jacquere_quota3[\"Target Gene Symbol\"])\n",
    "intended_target_symbols=jacquere_og[\"Target Gene Symbol\"].tolist()\n",
    "jacquere_quota3[\"All Jacquere target symbols that guide maps to\"]=jacquere_quota3.apply(lambda x: set(gene for gene in [x[\"Gene targeted\"]]+x[\"NCBI 08-2024 mapped gene symbols\"].split(\",\")+x[\"Ensembl.v113 mapped gene symbols\"].split(\",\") if gene in intended_target_symbols),axis=1)\n",
    "intended_target_IDs=jacquere_og[\"Input\"].tolist()\n",
    "jacquere_quota3[\"All Jacquere target IDs that guide maps to\"]=jacquere_quota3.apply(lambda x: set(gene for gene in [x[\"Input\"]]+x[\"NCBI 08-2024 mapped gene IDs\"].split(\",\")+x[\"Ensembl.v113 mapped gene IDs\"].split(\",\") if gene in intended_target_IDs),axis=1)\n",
    "\n",
    "jacquere_for_gene_counts= jacquere_quota3.explode(\"All Jacquere target IDs that guide maps to\")\n",
    "jacquere_for_gene_counts[\"Target ID\"]=jacquere_for_gene_counts[\"All Jacquere target IDs that guide maps to\"]\n",
    "\n",
    "print(\"# targeting guides in jacquere \", len(jacquere_for_gene_counts[\"sgRNA Sequence\"].unique()))\n",
    "print(\"# genes targeted\", len(jacquere_for_gene_counts[\"Input\"].unique()))\n",
    "\n",
    "num_guides_gene=pd.DataFrame(jacquere_for_gene_counts[\"Target ID\"].value_counts()).reset_index()\n",
    "num_guides_gene[\"# Guides targeting gene\"]=num_guides_gene[\"count\"]\n",
    "#4 is a placehold here for genes targeted by >3 guides\n",
    "num_guides_gene[\"# Guides targeting gene\"]=np.where(num_guides_gene[\"# Guides targeting gene\"]>3,4,num_guides_gene[\"# Guides targeting gene\"])\n",
    "num_guides_gene[\"Source\"]=np.where(num_guides_gene[\"Target ID\"].str.count(\"ENS\")>0, \"GENCODE\",\n",
    "                       np.where(num_guides_gene[\"Target ID\"].str.count(\"CHS\")>0,\"CHESS\",\"RefSeq\"))\n",
    "source_quota_counts=pd.DataFrame(num_guides_gene[[\"Source\",\"# Guides targeting gene\"]].value_counts()).reset_index()\n",
    "\n",
    "#add in genes targeted by 0 guides\n",
    "genes_0_jacquere_guides=jacquere_og[jacquere_og[\"Input\"].isin(jacquere_for_gene_counts[\"Target ID\"].tolist())==False][[\"Source\",\"Input\",\"Target Gene Symbol\"]].drop_duplicates()\n",
    "source_counts_0_genes=pd.DataFrame(genes_0_jacquere_guides[\"Source\"].value_counts()).reset_index()\n",
    "source_counts_0_genes[\"# Guides targeting gene\"]=0\n",
    "source_quota_counts_table=pd.concat([source_quota_counts,source_counts_0_genes]).pivot(index=\"# Guides targeting gene\",columns=\"Source\",values=\"count\").fillna(0).sort_values(by=\"# Guides targeting gene\",ascending=False).astype(int)\n",
    "source_quota_counts_table[\"Total\"]=source_quota_counts_table.sum(axis=1)\n",
    "source_quota_counts_table=source_quota_counts_table.rename(index={4:\"4+\"})\n",
    "\n",
    "#plot\n",
    "fig, ax = plt.subplots(figsize=(15,2))\n",
    "ax.table(cellText=source_quota_counts_table[[\"Total\",\"GENCODE\",\"RefSeq\",\"CHESS\"]].values, \n",
    "         colLabels=source_quota_counts_table[[\"Total\",\"GENCODE\",\"RefSeq\",\"CHESS\"]].columns, \n",
    "         loc='center',colColours=['lightblue','0.8','0.8','0.8'],colWidths=[.1,.1,.1,.1],\n",
    "         rowLabels=[\"# Targeted by >3 guides\",\"# Targeted by 3 guides\",\"# Targeted by 2 guides\",\"# Targeted by 1 guide\",\"# Not targetable\"])\n",
    "ax.axis('off')\n",
    "fig.tight_layout()\n",
    "plt.title(\"Genes targeted in Jacquere (subsequent catalogs report those not recognized by prior) \")\n",
    "gpp.savefig(\"../Figures/jacquere_gene_counts_source_quota.pdf\",bbox_inches=\"tight\",dpi=600)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecbabf1b-ea46-4ece-a7f4-26ffea6b8f52",
   "metadata": {},
   "source": [
    "Output list of untargetable genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0672e9-50ca-45bb-a227-94d0d5d8ac6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#genes_0_jacquere_guides.to_csv(\"../Jacquere_untargetable_genes.csv\",index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b00b646a-4318-4632-916f-db5b0ce303eb",
   "metadata": {},
   "source": [
    "Generate report in which guides are not repeated, and each line indicates all genes that the guide maps to. This is used to construct the Clone Pool Order to GPP Production for Jacquere "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c0b07e9-8f76-4e2a-8518-1cb74fb35c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "jacquere_no_guide_repeats=jacquere_for_gene_counts[[\"sgRNA Sequence\",\"Target ID\",\"All Jacquere target symbols that guide maps to\"]].groupby(\"sgRNA Sequence\").agg(list).reset_index()\n",
    "\n",
    "jacquere_no_guide_repeats[\"Target Gene ID\"]=jacquere_no_guide_repeats[\"Target ID\"].apply(lambda x:\"|\".join([str(i) for i in x]))\n",
    "jacquere_no_guide_repeats[\"Target Gene Symbol\"]=jacquere_no_guide_repeats[\"All Jacquere target symbols that guide maps to\"].apply(lambda x:\"|\".join([str(i) for i in x[0]]))\n",
    "\n",
    "#add in the negative controls\n",
    "jacquere_no_guide_repeats_export=pd.concat([jacquere_no_guide_repeats[[\"sgRNA Sequence\",\"Target Gene Symbol\",\"Target Gene ID\"]],intergenics[[\"sgRNA Sequence\",\"Target Gene Symbol\",\"Target Gene ID\"]],nontargeting[[\"sgRNA Sequence\",\"Target Gene Symbol\",\"Target Gene ID\"]]])\n",
    "\n",
    "#jacquere_no_guide_repeats_export.to_csv(\"../Jacquere_PerGuideAnnotations.csv\",index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77a630e3-d960-436c-bb51-8858cab3cff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "jacquere_guides=jacquere_quota3[\"sgRNA Sequence\"].tolist() #note: only includes targeting guides \n",
    "jacquere_guides_in_brunello=[g for g in jacquere_guides if g in brunello_guides]\n",
    "print(\"% of jacquere guides that are in brunello:\", 100*len(jacquere_guides_in_brunello)/len(jacquere_guides))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "549363b1-c822-47b7-8644-df8f25910d9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "v=venn3([set(brunello_guides),set(gattinara_guides),set(jacquere_guides)],set_labels=('Brunello (2016)','Gattinara (2018)','Jacquere (2025)'),\n",
    "        layout_algorithm=DefaultLayoutAlgorithm(),\n",
    "        set_colors= (\"y\",\"g\",\"b\")) \n",
    "v\n",
    "plt.title(\"Overlap of Targeting Guides in CRISPRko Cas9 Libraries\")\n",
    "gpp.savefig(\"../Figures/GPP_CRISPRko_cas9_genomewide_library_overlap.pdf\",bbox_inches=\"tight\",dpi=600)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d40fe7c8-b0e9-48b6-b106-7221f28ca584",
   "metadata": {},
   "source": [
    "### Identify if any guides violate gnomAD filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef9dd820-3ca9-4ae4-ba50-94c52b0db1e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"% of targeting genes in Jacquere that violate gnomAD ancestry filter\",100*len(jacquere_quota3[jacquere_quota3[\"Picking Notes\"].str.count(\"Ancestry\")>0])/len(jacquere_quota3))\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
